docs/proposals/20241127-namespace-separation.md

```diff
@@ -0,0 +1,297 @@
+* We need to select multiple namespaces using `--namespace <ns1>` ... `--namespace <nsN>` to watch multiple namespaces.
+
+##### FR2 - watch on all namespaces excluding multiple namespaces
+* We need to selected excluded namespaces using `--excluded-namespace <ens1>` ... `--namespace <ensN>` to watch on all namespaces excluding selected namespaces.
+
+#### Non-Functional Requirements
+
+We consider that all CAPI instances:
+- are using the same container image
+- are sharing CRDs and ClusterRole
+
+### Implementation Details/Notes/Constraints
+
+There is a hash `watchNamespaces` of `cache.Config{}` objects [here](https://github.com/kubernetes-sigs/cluster-api/pull/11397/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261L319-R319):
+```go
+  var watchNamespaces map[string]cache.Config
+  ...
+  ctrlOptions := ctrl.Options{
+      ...
+      Cache: cache.Options{
+          DefaultNamespaces: watchNamespaces,
+          ....
+      }
+```
+
+#### Current state:
+* when `watchNamespaces` == `nil` then all namespaces are watched
+* when `watchNamespaces` == `map[string]cache.Config{ watchNamespace: {} }` then only namespace specified by `watchNamespace` is watched
+
+#### Watch on multiple namespaces
+We are suggesting to [update](https://github.com/kubernetes-sigs/cluster-api/pull/11397/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R320-R323) `watchNamespace` to a `watchNamespacesList` list:
+```go
+   if watchNamespacesList != nil {
+        watchNamespaces = map[string]cache.Config{}
+        for _, watchNamespace := range watchNamespacesList {
+            watchNamespaces[watchNamespace] = cache.Config{}
+        }
+    }
+```
+then the namespaces contained in `watchNamespacesList` will be watched by the instance.
+
+#### Exclude watching on selected namespaces
+1. We are suggesting to [create the fieldSelector condition](https://github.com/kubernetes-sigs/cluster-api/pull/11370/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R331-R339) matching all namespaces excluding the list defined in `watchExcludedNamespaces`:
+   ```go
+    var fieldSelector fields.Selector
+    if watchExcludedNamespaces != nil {
+        var conditions []fields.Selector
+        for i := range watchExcludedNamespaces {
+            conditions = append(conditions, fields.OneTermNotEqualSelector("metadata.namespace", watchExcludedNamespaces[i]))
+        }
+        fieldSelector = fields.AndSelectors(conditions...)
+    }
+   ```
+2. Then we can use the `fieldSelector` to set the `DefaultFieldSelector` value:
+    ```go
+        ctrlOptions := ctrl.Options{
+            ...
+            Cache: cache.Options{
+                DefaultFieldSelector: fieldSelector,
+            ....
+        }
+    ```
+and then the namespaces contained in `watchNamespacesList` will be excluded from watching by the instance.
```

Comment:
What is the rbac implication of this field selector, do I not need RBAC on that namespace?